File: src/files.cc
#include <nan.h>
#include "files.h"
#include "lib.h"

using namespace Nan;

NAN_METHOD(AddFile) {
    PWChar filename = (PWChar) node::Buffer::Data(info[0]->ToObject());
    WordBool ignoreExists(info[1]->BooleanValue());
    PCardinal handle = (PCardinal) node::Buffer::Data(info[2]->ToObject());
    WordBool success = xelib.functions.AddFile(filename, ignoreExists, handle);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(FileByIndex) {
    Integer index = info[0]->Int32Value();
    PCardinal handle = (PCardinal) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.FileByIndex(index, handle);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(FileByLoadOrder) {
    Integer loadOrder = info[0]->Int32Value();
    PCardinal handle = (PCardinal) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.FileByLoadOrder(loadOrder, handle);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(FileByName) {
    PWChar filename = (PWChar) node::Buffer::Data(info[0]->ToObject());
    PCardinal handle = (PCardinal) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.FileByName(filename, handle);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(FileByAuthor) {
    PWChar author = (PWChar) node::Buffer::Data(info[0]->ToObject());
    PCardinal handle = (PCardinal) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.FileByAuthor(author, handle);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(NukeFile) {
    Cardinal handle = info[0]->Uint32Value();
    WordBool success = xelib.functions.NukeFile(handle);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(RenameFile) {
    Cardinal handle = info[0]->Uint32Value();
    PWChar filename = (PWChar) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.RenameFile(handle, filename);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(SaveFile) {
    Cardinal handle = info[0]->Uint32Value();
    PWChar filePath = (PWChar) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.SaveFile(handle, filePath);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetRecordCount) {
    Cardinal handle = info[0]->Uint32Value();
    PInteger count = (PInteger) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.GetRecordCount(handle, count);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetOverrideRecordCount) {
    Cardinal handle = info[0]->Uint32Value();
    PInteger count = (PInteger) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.GetOverrideRecordCount(handle, count);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(MD5Hash) {
    Cardinal handle = info[0]->Uint32Value();
    PInteger len = (PInteger) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.MD5Hash(handle, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(CRCHash) {
    Cardinal handle = info[0]->Uint32Value();
    PInteger len = (PInteger) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.CRCHash(handle, len);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(SortEditorIDs) {
    Cardinal handle = info[0]->Uint32Value();
    PWChar signature = (PWChar) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.SortEditorIDs(handle, signature);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(SortNames) {
    Cardinal handle = info[0]->Uint32Value();
    PWChar signature = (PWChar) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.SortNames(handle, signature);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

NAN_METHOD(GetFileLoadOrder) {
    Cardinal handle = info[0]->Uint32Value();
    PInteger loadOrder = (PInteger) node::Buffer::Data(info[1]->ToObject());
    WordBool success = xelib.functions.GetFileLoadOrder(handle, loadOrder);
    info.GetReturnValue().Set(Nan::New((bool) success));
}

File: src/js/files.js
module.exports = function(lib, xelib, helpers) {
    let { Fail, elementContext, GetHandle, GetString, GetInteger, wcb } = helpers;

    // FILE HANDLING METHODS
    Object.assign(xelib, {
        AddFile: function(filename, ignoreExists = false) {
            return GetHandle(function(_res) {
                if (!lib.AddFile(wcb(filename), ignoreExists, _res))
                    Fail(`Failed to add new file: ${filename}`);
            });
        },
        FileByIndex: function(index) {
            return GetHandle((_res) => lib.FileByIndex(index, _res));
        },
        FileByLoadOrder: function(loadOrder) {
            return GetHandle((_res) => lib.FileByLoadOrder(loadOrder, _res));
        },
        FileByName: function(filename) {
            return GetHandle((_res) => lib.FileByName(wcb(filename), _res));
        },
        FileByAuthor: function(author) {
            return GetHandle((_res) => lib.FileByAuthor(wcb(author), _res));
        },
        NukeFile: function(id) {
            if (!lib.NukeFile(id))
                Fail(`Failed to nuke file: ${id}`);
        },
        RenameFile: function(id, newFileName) {
            if (!lib.RenameFile(id, wcb(newFileName)))
                Fail(() => `Failed to rename file ${elementContext(id)} to ${newFileName}`);
        },
        SaveFile: function(id, filePath = '') {
            if (!lib.SaveFile(id, wcb(filePath)))
                Fail(() => `Failed to save file: ${elementContext(id)}`);
        },
        GetRecordCount: function(id) {
            return GetInteger(function(_res) {
                if (!lib.GetRecordCount(id, _res))
                    Fail(() => `Failed to get record count for: ${elementContext(id)}`);
            });
        },
        GetOverrideRecordCount: function(id) {
            return GetInteger(function(_res) {
                if (!lib.GetOverrideRecordCount(id, _res))
                    Fail(() => `Failed to get override record count for: ${elementContext(id)}`);
            });
        },
        MD5Hash: function(id) {
            return GetString(function(_len) {
                if (!lib.MD5Hash(id, _len))
                    Fail(() => `Failed to get MD5 Hash for: ${elementContext(id)}`);
            });
        },
        CRCHash: function(id) {
            return GetString(function(_len) {
                if (!lib.CRCHash(id, _len))
                    Fail(() => `Failed to get CRC Hash for: ${elementContext(id)}`);
            });
        },
        GetFileLoadOrder: function(id) {
            return GetInteger(function(_res) {
                if (!lib.GetFileLoadOrder(id, _res))
                    Fail(() => `Failed to file load order for: ${elementContext(id)}`);
            });
        },
        GetFileHeader: function(id) {
            return xelib.GetElement(id, 'File Header');
        },
        SortEditorIDs: function(id, sig) {
            if (!lib.SortEditorIDs(id, wcb(sig)))
                Fail(() => `Failed to sort ${sig} EditorIDs for: ${elementContext(id)}`);
        },
        SortNames: function(id, sig) {
            if (!lib.SortNames(id, wcb(sig)))
                Fail(() => `Failed to sort ${sig} Names for: ${elementContext(id)}`);
        }
    });
};
